### ISSUE TITLE
Peptide and Protein Quantification Experimentation

### ISSUE THREAD

**Issue by CCranney** [xomicsdatascience/CsoDIAq#64]
We are trying to find ways to optimize the accuracy of the ionCount variable in the csodiaq output. Specifically, we should expect peptides and proteins with an expected dilution of the same sample to be found in quantities proportional to that dilution level. We addressed this issue in [our 2022 paper](https://pubs.acs.org/doi/10.1021/acs.analchem.2c02249?goto=supporting-info), and had found an acceptable solution, but we are looking to improve the ion-count-to-quantity accuracy. I am creating this issue to track progress on this front.

**Comment by CCranney** [comment]
I am trying to obtain scripts used for the analysis in the paper. Until then, using the same dataset as was used in the paper and running it through the current dev branch output for common peptides, I wrote the following script to consolidate dilution series data.

```python
import pandas as pd

def determine_dilution_from_file_name(fileName):
    if 'hela1ug' in fileName:
        return 1.0
    elif 'hela05ug' in fileName:
        return 0.5
    elif 'hela025ug' in fileName:
        return 0.25
    elif 'hela0125ug' in fileName:
        return 0.125

commonPeptidesFile = '<path_to_common_peptides_file>'
df = pd.read_csv(commonPeptidesFile, index_col=0)
#df = df.drop('CsoDIAq-file_hela0125ug240k_20220305180559_corrected_fullOutput') # extra file removal
#df = df.drop('CsoDIAq-file_hela0125ug240k_corrected_fullOutput') # extra file removal
df = df.loc[:, df.ne(0).all()] # drop any peptide with any zero value
df['dilution'] = [determine_dilution_from_file_name(fileName) for fileName in df.index] # add dilution level to each row
dilutionDf = df.groupby(['dilution']).mean()
```

**Comment by CCranney** [comment]
In the above code, I have 2 variables I determined should be investigated - dropping one of the extra 0125ug files, and determining the mean of the triplicate values for each peptide (as opposed to the median), set on the last line. I normalized the data (see next comment). The results of this analysis are below.

```

<median of the triplicate>
no exclusion
            median      mean       std
dilution                              
0.125     0.274624  0.338407  0.760964
0.250     0.405623  0.497337  1.224005
0.500     0.545086  0.675209  1.733595
1.000     1.000000  1.000000  0.000000

exclude CsoDIAq-file_hela0125ug240k_20220305180559_corrected_fullOutput
            median      mean       std
dilution
0.125     0.280253  0.345566  0.788651
0.250     0.404862  0.495025  1.203881
0.500     0.543918  0.672786  1.705188
1.000     1.000000  1.000000  0.000000

exclude CsoDIAq-file_hela0125ug240k_corrected_fullOutput
            median      mean       std
dilution
0.125     0.268246  0.329459  0.707957
0.250     0.405623  0.496712  1.211183
0.500     0.545702  0.675208  1.715221
1.000     1.000000  1.000000  0.000000

exclude CsoDIAq-file_hela0125ug240k2_corrected_fullOutput
            median      mean       std
dilution
0.125     0.277998  0.342093  0.797912
0.250     0.405088  0.493873  1.207051
0.500     0.544220  0.671219  1.709645
1.000     1.000000  1.000000  0.000000

exclude CsoDIAq-file_hela0125ug240k3_corrected_fullOutput
            median      mean       std
dilution
0.125     0.265451  0.329190  0.716600
0.250     0.405167  0.494399  1.212326
0.500     0.544820  0.672021  1.716908
1.000     1.000000  1.000000  0.000000

*********************

<mean of the triplicate>
no exclusion
            median      mean       std
dilution
0.125     0.279779  0.348433  0.880628
0.250     0.412656  0.505217  1.367879
0.500     0.528994  0.666837  1.985957
1.000     1.000000  1.000000  0.000000

exclude CsoDIAq-file_hela0125ug240k_20220305180559_corrected_fullOutput
            median      mean       std
dilution
0.125     0.286741  0.356401  0.922884
0.250     0.412076  0.502198  1.344898
0.500     0.528550  0.664201  1.952841
1.000     1.000000  1.000000  0.000000

exclude CsoDIAq-file_hela0125ug240k_corrected_fullOutput
            median      mean       std
dilution
0.125     0.269603  0.328943  0.778181
0.250     0.412292  0.504297  1.353306
0.500     0.530576  0.666729  1.964599
1.000     1.000000  1.000000  0.000000

exclude CsoDIAq-file_hela0125ug240k2_corrected_fullOutput
            median      mean       std
dilution
0.125     0.285327  0.354923  0.902049
0.250     0.412057  0.501771  1.348873
0.500     0.528550  0.662808  1.958261
1.000     1.000000  1.000000  0.000000

exclude CsoDIAq-file_hela0125ug240k3_corrected_fullOutput
            median      mean       std
dilution
0.125     0.277280  0.345573  0.878958
0.250     0.412076  0.502085  1.354759
0.500     0.529088  0.663562  1.966710
1.000     1.000000  1.000000  0.000000
```

In general, I am planning on using the following baseline code for preprocessing data outputs (keeping all the files and using the median, not the mean, between files).

```python
import pandas as pd
import numpy as np

def determine_dilution_from_file_name(fileName):
    if 'hela1ug' in fileName:
        return 1.0
    elif 'hela05ug' in fileName:
        return 0.5
    elif 'hela025ug' in fileName:
        return 0.25
    elif 'hela0125ug' in fileName:
        return 0.125

commonPeptidesFile = '<path_to_common_peptides_file>'
df = pd.read_csv(commonPeptidesFile, index_col=0)
df = df.loc[:, df.ne(0).all()] # drop any peptide with any zero value
df['dilution'] = [determine_dilution_from_file_name(fileName) for fileName in df.index] # add dilution level to each row
dilutionDf = df.groupby(['dilution']).median()
```

**Comment by CCranney** [comment]
I explored ways in which to normalize the data for comparison. It appears all dilution levels were divided by the 1.0 dilution in the original analysis (@AlexandreHutton please confirm when you're back - see figures 4E-F). I did this for the dataset, then found the mean, median and standard deviation across dilution rows for comparison with other methods.

```python
normalizedDf = dilutionDf.div(dilutionDf.iloc[-1])
normalizedDf['median'] = normalizedDf.median(axis=1)
normalizedDf['mean'] = normalizedDf.mean(axis=1)
normalizedDf['std'] = normalizedDf.std(axis=1)
print('\nregular normalization')
print(normalizedDf.iloc[:, -3:])
```

which generates the following output:
```
regular normalization
            median      mean       std
dilution                              
0.125     0.274624  0.338407  0.760964
0.250     0.405623  0.497337  1.224005
0.500     0.545086  0.675209  1.733595
1.000     1.000000  1.000000  0.000000
```

**Comment by CCranney** [comment]
I explored an alternative method for normalizing the data. My thought process is that the ionCount variable is not strictly linear - it could be that doubling peptide quantity does not equate to doubling the ionCount. I'd imagine there's likely a logarithmic relationship at play, or something similar. In this method I took each dilution level, then subtracted and divided by the lowest dilution level. I used the following code:

```python
## (x-smallest)/smallest normalization
diffDf = dilutionDf - dilutionDf.loc[0.125].values.squeeze()
proportionDf = diffDf.div(dilutionDf.loc[0.125].values.squeeze())
proportionDf['median'] = proportionDf.median(axis=1)
proportionDf['mean'] = proportionDf.mean(axis=1)
proportionDf['std'] = proportionDf.std(axis=1)
print('\nproportional normalization')
print(proportionDf.iloc[:, -3:])
```

which generates the following output:
```
proportional normalization
            median      mean       std
dilution                              
0.125     0.000000  0.000000  0.000000
0.250     0.492977  0.531392  0.394718
0.500     0.973479  1.162212  0.886242
1.000     2.641343  3.403132  3.168924
```

The standard deviation across dilution levels seems too wildly large to use this normalization method.

**Comment by CCranney** [comment]
I also investigated methods for calculating correlation - when playing with the various methods for ionCount calculation, we'd want to find a method that is most tightly correlated to dilution levels. I used pearson and spearman correlation tests with the following code (note that p-value significance is indicated with `*` symbols):

```python
from scipy.stats import pearsonr, spearmanr
# Correlation Tests
transposedDf = dilutionDf.T
## Pearson correlation
rho = transposedDf.corr()
pval = transposedDf.corr(method=lambda x, y: pearsonr(x, y)[1]) - np.eye(*rho.shape)
p = pval.applymap(lambda x: ''.join(['*' for t in [.05, .01, .001] if x<=t]))
pearson = rho.round(5).astype(str) + p
print('\npearson')
print(pearson)

## Spearman correlation
rho = transposedDf.corr(method='spearman')
pval = transposedDf.corr(method=lambda x, y: spearmanr(x, y)[1]) - np.eye(*rho.shape)
p = pval.applymap(lambda x: ''.join(['*' for t in [.05, .01, .001] if x<=t]))
spearman = rho.round(5).astype(str) + p
print('\nspearman')
print(spearman)
```

which generates the following output:
```
pearson
dilution       0.125       0.250       0.500       1.000
dilution                                                
0.125         1.0***  0.98988***  0.99366***  0.96225***
0.250     0.98988***      1.0***  0.99776***  0.97438***
0.500     0.99366***  0.99776***      1.0***  0.97296***
1.000     0.96225***  0.97438***  0.97296***      1.0***

spearman
dilution       0.125       0.250       0.500       1.000
dilution                                                
0.125         1.0***  0.97393***  0.94654***  0.85331***
0.250     0.97393***      1.0***  0.97421***  0.87888***
0.500     0.94654***  0.97421***      1.0***  0.89302***
1.000     0.85331***  0.87888***  0.89302***      1.0***
```

It appears the pearson correlation has a tighter positive correlation generally.

**Comment by CCranney** [comment]
Re-evaluated the above reproach to better reflect what was done in the paper. In the paper, no mean/median is taken of triplicate. Instead, a pearson and/or spearman correlation is taken of each peptide quantity. On a graph, this peptide quantity would be the y-axis, while the dilution would be the x-axis, as in the following basic example (taken from the dataset for the `(UniMod:5)ADKPDMGEIASFDK` peptide):

![Figure_1](https://github.com/xomicsdatascience/CsoDIAq/assets/11773171/a535cd67-9e6d-4868-815e-4cec06540d60)

A pearson/spearman correlation would then be taken for each value for a given peptide. Anything with a negative correlation was removed (as determined by a negative pearson coefficient). We also performed a Benjamini-Hochberg correction to create an adjusted p-value. I am still researching, but I don't believe that correction was applied in the original paper.

```python
import pandas as pd
import numpy as np
from scipy.stats import pearsonr, spearmanr, false_discovery_control, linregress
pd.set_option('display.max_columns', None)

def determine_dilution_from_file_name(fileName):
    if 'hela1ug' in fileName:
        return 1.0
    elif 'hela05ug' in fileName:
        return 0.5
    elif 'hela025ug' in fileName:
        return 0.25
    elif 'hela0125ug' in fileName:
        return 0.125

commonPeptidesFile = '/Users/cranneyc/Desktop/csodiaqOutput/currentDevBranchOutput/2023_08_29_dilutionSeries/fdrScores-macc-average/commonPeptides.csv'
df = pd.read_csv(commonPeptidesFile, index_col=0)
df = df.drop('CsoDIAq-file_hela0125ug240k_20220305180559_corrected_fullOutput')
df = df.loc[:, df.ne(0).all()] # drop any peptide with any zero value
dilutions = [determine_dilution_from_file_name(fileName) for fileName in df.index] # extract dilution level from file name

# Correlation Tests
transposedDf = df.T
correlationDf = transposedDf.copy()
correlationDf['slope'] = transposedDf.apply(lambda peptide: linregress(np.array(peptide) / max(peptide), dilutions).slope, axis=1)
correlationDf['pearsonr'] = transposedDf.apply(lambda peptide: pearsonr(list(peptide), dilutions)[0], axis=1)
correlationDf['pearsonPValue'] = transposedDf.apply(lambda peptide: pearsonr(list(peptide), dilutions)[1], axis=1)
correlationDf = correlationDf[correlationDf['pearsonr']>0]
correlationDf['pearsonPValueAdjusted'] = false_discovery_control(correlationDf['pearsonPValue'], method='bh')
correlationDf['spearmanr'] = transposedDf.apply(lambda peptide: spearmanr(list(peptide), dilutions)[0], axis=1)
correlationDf['spearmanPValue'] = transposedDf.apply(lambda peptide: spearmanr(list(peptide), dilutions)[1], axis=1)
correlationDf['spearmanPValueAdjusted'] = false_discovery_control(correlationDf['spearmanPValue'], method='bh')
print(correlationDf.iloc[:, -7]) # note: peptide quantity was normalized by dividing by the largest ionCount value to adjust the expected slope to 1
print(correlationDf.iloc[:, -6:-3])
print(correlationDf.iloc[:, -3:])
print(f'\ntotal number of peptides: {len(correlationDf)}')
print(f'(pearson) Number of peptides below 0.01 adjusted p-value: {len(correlationDf[correlationDf["pearsonPValueAdjusted"]<0.01])}')
print(f'(spearman) Number of peptides below 0.01 adjusted p-value: {len(correlationDf[correlationDf["spearmanPValueAdjusted"]<0.01])}')
```

The output of running the above was:
```
(UniMod:5)ADKPDMGEIASFDK      1.232014
(UniMod:5)SAHLQWMVVR          1.110466
(UniMod:5)SDKPDMAEIEK         1.310796
(UniMod:5)SSGIHVALVTGGNK      1.157547
(UniMod:5)TTNAGPLHPYWPQHLR    1.304495
                                ...   
YVEVSPATER                    1.181675
YVIHTVGPIAYGEPSASQAAELR       1.209329
YVMTTTTLER                    1.245333
YWLC(UniMod:4)AATGPSIK        1.263024
YYTSASGDEMVSLK                1.351250
Name: slope, Length: 1976, dtype: float64
                            pearsonr  pearsonPValue  pearsonPValueAdjusted
(UniMod:5)ADKPDMGEIASFDK    0.989686   9.036512e-10           1.004196e-07
(UniMod:5)SAHLQWMVVR        0.982230   1.354385e-08           2.152563e-07
(UniMod:5)SDKPDMAEIEK       0.969300   2.039759e-07           8.075725e-07
(UniMod:5)SSGIHVALVTGGNK    0.930709   1.118909e-05           1.583786e-05
(UniMod:5)TTNAGPLHPYWPQHLR  0.963554   4.763728e-07           1.374179e-06
...                              ...            ...                    ...
YVEVSPATER                  0.947972   2.750319e-06           4.954085e-06
YVIHTVGPIAYGEPSASQAAELR     0.958372   9.178747e-07           2.166768e-06
YVMTTTTLER                  0.973590   9.680998e-08           5.319623e-07
YWLC(UniMod:4)AATGPSIK      0.974474   8.177724e-08           4.926580e-07
YYTSASGDEMVSLK              0.951995   1.851952e-06           3.626816e-06

[1976 rows x 3 columns]
                            spearmanr  spearmanPValue  spearmanPValueAdjusted
(UniMod:5)ADKPDMGEIASFDK     0.971625    1.381174e-07            5.569795e-07
(UniMod:5)SAHLQWMVVR         0.971625    1.381174e-07            5.569795e-07
(UniMod:5)SDKPDMAEIEK        0.950034    2.254862e-06            5.604539e-06
(UniMod:5)SSGIHVALVTGGNK     0.906850    4.715314e-05            7.064034e-05
(UniMod:5)TTNAGPLHPYWPQHLR   0.906850    4.715314e-05            7.064034e-05
...                               ...             ...                     ...
YVEVSPATER                   0.820484    1.076979e-03            1.275846e-03
YVIHTVGPIAYGEPSASQAAELR      0.863667    2.937192e-04            3.768761e-04
YVMTTTTLER                   0.928442    1.309263e-05            2.391039e-05
YWLC(UniMod:4)AATGPSIK       0.950034    2.254862e-06            5.604539e-06
YYTSASGDEMVSLK               0.950034    2.254862e-06            5.604539e-06

[1976 rows x 3 columns]

total number of peptides: 1976
(pearson) Number of peptides below 0.01 adjusted p-value: 1843
(spearman) Number of peptides below 0.01 adjusted p-value: 1798
```

My results are slightly better than what was reported in the paper. It may be one of my 0.125 dilution files is wrong - I have 4 and excluded the one with an outlier name. I am researching which 3 were used in the paper in case a different file was excluded. I will run everything through the main branch (rather than the dev branch) and see if the updated code has different outputs to see if the differences extend beyond that extra datapoint issue.

original paper results:
```
total number of peptides: 2321
Number of peptides below **0.05** NON-adjusted p-value for pearson correlation: 1799
```

**Comment by CCranney** [comment]
I ran the common_peptide.csv output of the main branch through the above code and got the following results. Again, they differ from what is in the paper, something to investigate when I get the original analysis methods.

```
peptide
VAGMDVELTVEER                     1.140971
ALVDHENVISC(UniMod:4)PHLGASTK     1.325572
SLTNDWEDHLAVK                     1.004530
IYPHGLVLLDLQSYDGDAQGK             1.103378
THFFLNAGNLC(UniMod:4)NLNYGEGPK    1.321087
                                    ...   
TEVNSGFFYK                        1.212208
WLAIDANAR                         1.214459
DGYNYTLSK                         0.439598
DIPGLTDTTVPR                      1.278914
TFVLEVMGR                         1.001312
Name: slope, Length: 1732, dtype: float64
                                pearsonr  pearsonPValue  pearsonPValueAdjusted
peptide                                                                       
VAGMDVELTVEER                   0.957160   1.057365e-06           3.341890e-06
ALVDHENVISC(UniMod:4)PHLGASTK   0.939440   5.791876e-06           1.187163e-05
SLTNDWEDHLAVK                   0.961968   5.878695e-07           2.237780e-06
IYPHGLVLLDLQSYDGDAQGK           0.992446   1.912128e-10           5.519675e-08
THFFLNAGNLC(UniMod:4)NLNYGEGPK  0.970645   1.634046e-07           1.021721e-06
...                                  ...            ...                    ...
TEVNSGFFYK                      0.901091   6.301853e-05           9.103881e-05
WLAIDANAR                       0.918763   2.428239e-05           3.923236e-05
DGYNYTLSK                       0.288011   3.639896e-01           3.706231e-01
DIPGLTDTTVPR                    0.946372   3.191426e-06           7.592788e-06
TFVLEVMGR                       0.912087   3.562896e-05           5.483950e-05

[1732 rows x 3 columns]
                                spearmanr  spearmanPValue  \
peptide                                                     
VAGMDVELTVEER                    0.928442    1.309263e-05   
ALVDHENVISC(UniMod:4)PHLGASTK    0.885259    1.288056e-04   
SLTNDWEDHLAVK                    0.928442    1.309263e-05   
IYPHGLVLLDLQSYDGDAQGK            0.971625    1.381174e-07   
THFFLNAGNLC(UniMod:4)NLNYGEGPK   0.971625    1.381174e-07   
...                                   ...             ...   
TEVNSGFFYK                       0.885259    1.288056e-04   
WLAIDANAR                        0.971625    1.381174e-07   
DGYNYTLSK                        0.367058    2.405170e-01   
DIPGLTDTTVPR                     0.906850    4.715314e-05   
TFVLEVMGR                        0.928442    1.309263e-05   

                                spearmanPValueAdjusted  
peptide                                                 
VAGMDVELTVEER                             2.877721e-05  
ALVDHENVISC(UniMod:4)PHLGASTK             2.039225e-04  
SLTNDWEDHLAVK                             2.877721e-05  
IYPHGLVLLDLQSYDGDAQGK                     6.913853e-07  
THFFLNAGNLC(UniMod:4)NLNYGEGPK            6.913853e-07  
...                                                ...  
TEVNSGFFYK                                2.039225e-04  
WLAIDANAR                                 6.913853e-07  
DGYNYTLSK                                 2.440395e-01  
DIPGLTDTTVPR                              8.282885e-05  
TFVLEVMGR                                 2.877721e-05  

[1732 rows x 3 columns]

total number of peptides: 1732
(pearson) Number of peptides below 0.01 adjusted p-value: 1579
(spearman) Number of peptides below 0.01 adjusted p-value: 1524
```

**Comment by CCranney** [comment]
Using the same code, I analyzed the `commonProteins.csv` output. This is for my original averaging algorithm that was not tested on actual data until now.

```
total number of proteins: 914
(pearson) Number of proteins below 0.01 adjusted p-value: 850
(spearman) Number of proteins below 0.01 adjusted p-value: 808
```

I also implemented a new algorithm based on the maxLFQ algorithm described in [this paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4159666/). In addition to the paper description I also based my code on an R and C++ implementation in [this github repository](https://github.com/vdemichev/diann-rpackage/tree/master). My initial testing suggests that it accurately depicts intensity differences between samples.

```
total number of proteins: 1159
(pearson) Number of proteins below 0.01 adjusted p-value: 1089
(spearman) Number of proteins below 0.01 adjusted p-value: 1090
```

Note that the number of overall proteins increased. That is because proteins identified in one sample are evaluated across all samples, including those in which it was not originally identified. If they share peptides that correspond to that protein, it is possible for the protein to be re-identified across other samples.

**Comment by CCranney** [comment]
Going back to peptide experimentation, I tried just using the most intense query peak for a spectral match as the ionCount variable. The results were:

```
total number of peptides: 1987
(pearson) Number of peptides below 0.01 adjusted p-value: 1859
(spearman) Number of peptides below 0.01 adjusted p-value: 1829
```

A few more peptides were included - from my research it looks like these are primarily because a small number of peptides have no m/z value matches above the precursor m/z. It also performed slightly better, though that difference is marginal at best.

I also tried summing the three most intense peaks. In general, these results were significantly worse, but I'm recording it here for the sake of future review.

```
total number of peptides: 1986
(pearson) Number of peptides below 0.01 adjusted p-value: 1766
(spearman) Number of peptides below 0.01 adjusted p-value: 1659
```

**Comment by CCranney** [comment]
I realized while writing system tests for the maxlfq function that I was including too many proteins. In addition to the extras addressed earlier where proteins excluded by idPicker were re-included (now updated: set to 0 quantity in `commonProteins.csv`), I was failing to exclude proteins in samples with less than 2 peptide matches with other samples. This is the newly-updated output. Lower quantity, but higher quality (97-98% quantified correctly).

```
total number of proteins: 609
(pearson) Number of proteins below 0.01 adjusted p-value: 595
(spearman) Number of proteins below 0.01 adjusted p-value: 598
```

**Comment by AlexandreHutton** [comment]
> I explored ways in which to normalize the data for comparison. It appears all dilution levels were divided by the 1.0 dilution in the original analysis (@AlexandreHutton please confirm when you're back - see figures 4E-F). I did this for the dataset, then found the mean, median and standard deviation across dilution rows for comparison with other methods.
  
Confirmed that the undiluted sample was used as a common reference.

**Comment by CCranney** [comment]
MaxLFQ implemented with zodiaq 2.0, closing this issue.
